source/extensions/filters/http/jwt_authn/extractor.cc

```diff
@@ -0,0 +1,154 @@
+ */
+struct JwtConstValueStruct {
+  // The header value prefix for Authorization.
+  const std::string BearerPrefix{"Bearer "};
+
+  // The default query parameter name to extract JWT token
+  const std::string AccessTokenParam{"access_token"};
+};
+typedef ConstSingleton<JwtConstValueStruct> JwtConstValues;
+
+// A base JwtLocation object to store token and specified_issuers.
+class JwtLocationBase : public JwtLocation {
+public:
+  JwtLocationBase(const std::string& token, const std::unordered_set<std::string>& issuers)
+      : token_(token), specified_issuers_(issuers) {}
+
+  // Get the token string
+  const std::string& token() const override { return token_; }
+
+  // Check if an issuer has specified the location.
+  bool isIssuerSpecified(const std::string& issuer) const override {
+    return specified_issuers_.find(issuer) != specified_issuers_.end();
+  }
+
+private:
+  // Extracted token.
+  std::string token_;
+  // Stored issuers specified the location.
+  const std::unordered_set<std::string>& specified_issuers_;
+};
+
+// The JwtLocation for header extraction.
+class JwtHeaderLocation : public JwtLocationBase {
+public:
+  JwtHeaderLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                    const LowerCaseString& header)
+      : JwtLocationBase(token, issuers), header_(header) {}
+
+  void removeJwt(Http::HeaderMap& headers) const override { headers.remove(header_); }
+
+private:
+  // the header name the JWT is extracted from.
+  const LowerCaseString& header_;
+};
+
+// The JwtLocation for param extraction.
+class JwtParamLocation : public JwtLocationBase {
+public:
+  JwtParamLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                   const std::string&)
+      : JwtLocationBase(token, issuers) {}
+
+  void removeJwt(Http::HeaderMap&) const override {
+    // TODO(qiwzhang): remove JWT from parameter.
+  }
+
+private:
+};
+
+} // namespace
+
+Extractor::Extractor(const JwtAuthentication& config) {
+  for (const auto& rule : config.rules()) {
+    for (const auto& header : rule.from_headers()) {
+      addHeaderConfig(rule.issuer(), LowerCaseString(header.name()), header.value_prefix());
+    }
+    for (const std::string& param : rule.from_params()) {
+      addParamConfig(rule.issuer(), param);
+    }
+
+    // If not specified, use default locations.
+    if (rule.from_headers().empty() && rule.from_params().empty()) {
+      addHeaderConfig(rule.issuer(), Http::Headers::get().Authorization,
+                      JwtConstValues::get().BearerPrefix);
+      addParamConfig(rule.issuer(), JwtConstValues::get().AccessTokenParam);
+    }
+  }
+}
+
+void Extractor::addHeaderConfig(const std::string& issuer, const LowerCaseString& header_name,
+                                const std::string& value_prefix) {
+  std::string map_key = header_name.get() + value_prefix;
+  auto& map_value = header_maps_[map_key];
+  if (!map_value) {
```

Comment:
Should we raise an exception and reject the config if there are dupes header locations, or do you prefer to just silently ignore later ones because it doesn't matter (or does it, for ordering reasons?!)?

Reply:
We should support different issuers with same header/prefix.  

For example,  two issuers, issuer1 and issuer2 can extract their token from "Authorization: Bearer".   

A request with headers:
Authroization: Bearer <issuer1_token>

it is OK, it will use issuer1 config to process.

Similarly,  it should work for:
Authroization: Bearer <issuer2_token>